src/containerapp/azext_containerapp/_utils.py

```diff
@@ -12,145 +12,85 @@
+from ._models import (ContainerAppCustomDomainEnvelope as ContainerAppCustomDomainEnvelopeModel)
 
 logger = get_logger(__name__)
 
 
+def register_provider_if_needed(cmd, rp_name):
+    if not _is_resource_provider_registered(cmd, rp_name):
+        _register_resource_provider(cmd, rp_name)
+
+
 def validate_container_app_name(name):
     if name and len(name) > MAXIMUM_CONTAINER_APP_NAME_LENGTH:
         raise ValidationError(f"Container App names cannot be longer than {MAXIMUM_CONTAINER_APP_NAME_LENGTH}. "
                               f"Please shorten {name}")
 
 
-# original implementation at azure.cli.command_modules.role.custom.create_service_principal_for_rbac
-# reimplemented to remove incorrect warning statements
-def create_service_principal_for_rbac(  # pylint:disable=too-many-statements,too-many-locals, too-many-branches, unused-argument, inconsistent-return-statements
-        cmd, name=None, years=None, create_cert=False, cert=None, scopes=None, role=None,
-        show_auth_for_sdk=None, skip_assignment=False, keyvault=None):
-    from azure.cli.command_modules.role.custom import (_graph_client_factory, TZ_UTC, _process_service_principal_creds,
-                                                       _validate_app_dates, create_application,
-                                                       _create_service_principal, _create_role_assignment,
-                                                       _error_caused_by_role_assignment_exists)
-
-    if role and not scopes or not role and scopes:
-        raise ArgumentUsageError("Usage error: To create role assignments, specify both --role and --scopes.")
-
-    graph_client = _graph_client_factory(cmd.cli_ctx)
-
-    years = years or 1
-    _RETRY_TIMES = 36
-    existing_sps = None
+def retry_until_success(operation, err_txt, retry_limit, *args, **kwargs):
+    try:
+        return operation(*args, **kwargs)
+    except Exception as e:
+        retry_limit -= 1
+        if retry_limit <= 0:
+            raise CLIInternalError(err_txt) from e
+        time.sleep(5)
+        logger.info(f"Encountered error: {e}. Retrying...")
+
+
+def get_vnet_location(cmd, subnet_resource_id):
+    parsed_rid = parse_resource_id(subnet_resource_id)
+    vnet_client = network_client_factory(cmd.cli_ctx)
+    location = vnet_client.virtual_networks.get(resource_group_name=parsed_rid.get("resource_group"),
+                                                virtual_network_name=parsed_rid.get("name")).location
+    return _normalize_location(cmd, location)
+
+
+def create_service_principal_for_github_action(cmd, scopes=None, role="contributor"):
+    from azure.cli.command_modules.role.custom import (create_application, create_service_principal,
+                                                       create_role_assignment, show_service_principal)
+    from azure.cli.command_modules.role.msgrpah._graph_client import GraphClient
```

Comment:
Please don't import from private module. Consider using `azure.cli.command_modules.role.graph_client_factory`: https://github.com/Azure/azure-cli/blob/cac3f06cf483d8c81f93c9cd556116e27aff2182/src/azure-cli/azure/cli/command_modules/role/__init__.py#L44

Reply:
Sounds good; just pushed this change